tools/codegen/gen_backend_stubs.py

```diff
@@ -40,34 +46,84 @@ def parse_backend_yaml(
     supported_autograd = yaml_values.pop('autograd', [])
     assert isinstance(supported, list), f'expected "autograd" to be a list, but got: {supported_autograd}'
 
+    if len(supported_autograd) > 0:
+        autograd_key = DispatchKey.try_parse(f'Autograd{backend}')
+        assert autograd_key is not None, f'The "autograd" key was specified, which indicates that you would like to override \
+the behavior of autograd for some operators on your backend. However "Autograd{backend}" is not a valid DispatchKey. \
+The set of valid dispatch keys is: {", ".join(k for k, v in DispatchKey.__members__.items())}'
+
     assert len(yaml_values.keys()) == 0, \
         f'{backend_yaml_path} contains unexpected keys: {", ".join(yaml_values.keys())}. \
 Only the following keys are supported: {", ".join(valid_keys)}'
 
     metadata: Dict[OperatorName, ExternalBackendMetadata] = {}
     for op in supported:
         op_name = OperatorName.parse(op)
-        m = ExternalBackendMetadata(op_name, backend, is_autograd=False)
+        m = ExternalBackendMetadata(op_name, is_autograd=False)
         metadata[m.operator] = m
     for op in supported_autograd:
         op_name = OperatorName.parse(op)
-        m = ExternalBackendMetadata(op_name, backend, is_autograd=True)
+        m = ExternalBackendMetadata(op_name, is_autograd=True)
         metadata[m.operator] = m
 
     native_functions_map: Dict[OperatorName, NativeFunction] = {
         f.func.name: f
         for f in concatMap(lambda f: [f] if isinstance(f, NativeFunction) else list(f.functions()), grouped_native_functions)
     }
 
+    def kernel_name(func: FunctionSchema) -> str:
+        # For external backends, we enforce that their names and signatures match the dispatcher convention
+        return dispatcher.name(func)
+
     def native_to_external(
             g: Union[NativeFunction, NativeFunctionsGroup]
     ) -> Union[ExternalBackendFunction, ExternalBackendFunctionsGroup]:
         if isinstance(g, NativeFunction):
             f = g
             m = metadata.get(f.func.name, None)
-            return ExternalBackendFunction(f, m)
+            dispatch_key = DispatchKey.parse(f'Autograd{backend}') \
+                if m is not None and m.is_autograd else DispatchKey.parse(backend)
+            kernel = kernel_name(f.func)
+            return ExternalBackendFunction(NativeFunction.with_dispatch_entry(f, dispatch_key, kernel), dispatch_key, m)
```

Comment:
I'm thinking back to https://github.com/pytorch/pytorch/pull/55050/files#r616215579

The stated motivation for patching in the dispatch information is to make it easier to reuse existing code in the codegen. This is very reasonable. But to me this more and more bends the orientation of the system towards trying to reuse the existing data structures or directly adding in the information as you need, and then augmenting it with side data (so the initial model doesn't have to changed).

Reply:
After talking to Ed, I'm working on a new patch that will change most of the contents of this PR:

Rather than augmenting NativeFunction directly to be external-backend-aware, I'm going to move data on NativeFunction that is backend-dependent into a different struct. Important stuff it'll contain includes (a) the kernel name (different per backend), and (b) whether the kernel is structured (technically the same for all in-tree backends, but can be different per external backend).